### PR TITLE
[Merged by Bors] - feat(CategoryTheory/ConnectedComponents) : extracted a definition from Cat to be used by categories in general

### PR DESCRIPTION
**Pull request by nrolland**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/ConnectedComponents.lean [hunk 1]
```diff
@@ -36,6 +36,12 @@ variable {J : Type u₁} [Category.{v₁} J]
 def ConnectedComponents (J : Type u₁) [Category.{v₁} J] : Type u₁ :=
   Quotient (Zigzag.setoid J)
 
+/-- Functors are mapped to functions from the connected components of their domain and target -/
+def Functor.mapConnectedComponents  {K : Type u₂} [Category.{v₂} K] (F : J ⥤ K) :
+    ConnectedComponents J → ConnectedComponents K :=
+  Quotient.lift (Quotient.mk (Zigzag.setoid _) ∘ F.obj)
+    (fun _ _ ↦ Quot.sound ∘ zigzag_obj_of_zigzag F)
```
Comment (dagurtomas) [inline Mathlib/CategoryTheory/ConnectedComponents.lean:43]: Declarations usually don't have type `A → B`. Instead, you should take an argument of type `ConnectedComponents J` and return a `ConnectedComponents K`.

File: Mathlib/CategoryTheory/ConnectedComponents.lean [hunk 2]
```diff
@@ -36,6 +36,12 @@ variable {J : Type u₁} [Category.{v₁} J]
 def ConnectedComponents (J : Type u₁) [Category.{v₁} J] : Type u₁ :=
   Quotient (Zigzag.setoid J)
 
+/-- Functors are mapped to functions from the connected components of their domain and target -/
```
Comment (joelriou) [inline Mathlib/CategoryTheory/ConnectedComponents.lean:39]:
```suggestion
/-- The map `ConnectedComponents J → ConnectedComponents K` induced by a functor `J ⥤ K`. -/
```

File: Mathlib/CategoryTheory/ConnectedComponents.lean [hunk 3]
```diff
@@ -36,6 +36,12 @@ variable {J : Type u₁} [Category.{v₁} J]
...
+def Functor.mapConnectedComponents  {K : Type u₂} [Category.{v₂} K] (F : J ⥤ K)
```
Comment (joelriou) [inline Mathlib/CategoryTheory/ConnectedComponents.lean:40]:
```suggestion
def Functor.mapConnectedComponents {K : Type u₂} [Category.{v₂} K] (F : J ⥤ K)
```

**Comment by joelriou** [comment]
It could be useful to add a (rfl) equational lemma for `Functor.mapConnectedComponents` which would assert that the image of the connected component of an object is the connected component of the image of this object.

File: Mathlib/CategoryTheory/ConnectedComponents.lean [hunk 4]
```diff
@@ -36,6 +36,17 @@ variable {J : Type u₁} [Category.{v₁} J]
...
+/-- The map `ConnectedComponents J → ConnectedComponents K` induced by a functor `J ⥤ K`. -/
+def Functor.mapConnectedComponents {K : Type u₂} [Category.{v₂} K] (F : J ⥤ K)
+    (x : ConnectedComponents J) : ConnectedComponents K :=
+  x |> Quotient.lift (Quotient.mk (Zigzag.setoid _) ∘ F.obj)
+    (fun _ _ ↦ Quot.sound ∘ zigzag_obj_of_zigzag F)
+
+@[simp]
+lemma image_of_cc_is_cc_of_image {K : Type u₂} [Category.{v₂} K] (F : J ⥤ K) :
+    F.mapConnectedComponents ∘ (Quotient.mk (Zigzag.setoid J))
+      =  Quotient.mk (Zigzag.setoid K) ∘ F.obj := rfl
```
Comment (joelriou) [inline Mathlib/CategoryTheory/ConnectedComponents.lean:48]: I was thinking of this statement:
```suggestion
@[simp]
lemma Functor.mapConnectedComponents_mk {K : Type u₂} [Category.{v₂} K] (F : J ⥤ K) (j : J) :
    F.mapConnectedComponents (Quotient.mk _ j) = Quotient.mk _ (F.obj j) := rfl
```
(Note also that the names of the lemmas are important. Also, according to mathlib's coding style, infix operators like `=` should not appear after a linebreak: they must appear before.)

**Comment by joelriou** [comment]
Thanks!

bors merge

**Pull request closed without merging**